lib/auth/keystore/gcp_kms.go

```diff
@@ -0,0 +1,454 @@
+	}
+	return true, nil
+}
+
+// DeleteUnusedKeys deletes all keys from KMS if they are:
+// 1. Labeled by this server (matching HostUUID) when they were created
+// 2. Not included in the argument usedKeys
+func (g *gcpKMSKeyStore) DeleteUnusedKeys(ctx context.Context, usedKeys [][]byte) error {
+	usedKmsKeyVersions := make(map[string]struct{})
+	for _, usedKey := range usedKeys {
+		keyID, err := parseGCPKMSKeyID(usedKey)
+		if err != nil {
+			// could be a different type of key
+			continue
+		}
+		usedKmsKeyVersions[keyID.keyVersionName] = struct{}{}
+	}
+
+	var unusedKeyIDs []gcpKMSKeyID
+
+	listKeyRequest := &kmspb.ListCryptoKeysRequest{
+		Parent: g.keyRing,
+		Filter: fmt.Sprintf("labels.%s=%s", hostLabel, g.hostUUID),
+	}
+	iter := g.kmsClient.ListCryptoKeys(ctx, listKeyRequest)
+	key, err := iter.Next()
+	for err == nil {
+		keyVersionName := key.Name + keyVersionSuffix
+		if _, used := usedKmsKeyVersions[keyVersionName]; !used {
+			unusedKeyIDs = append(unusedKeyIDs, gcpKMSKeyID{
+				keyVersionName: keyVersionName,
+			})
+		}
+		key, err = iter.Next()
+	}
+	if err != nil && !errors.Is(err, iterator.Done) {
+		return trace.Wrap(err)
+	}
+
+	for _, unusedKey := range unusedKeyIDs {
+		g.log.WithFields(logrus.Fields{"key_version": unusedKey.keyVersionName}).Info("deleting unused GCP KMS key created by this server")
+		err := g.deleteKey(ctx, unusedKey.marshal())
+		// Ignore errors where we can't destroy because the state is already
+		// DESTROYED or DESTROY_SCHEDULED
+		if err != nil && !strings.Contains(err.Error(), "has value DESTROY") {
+			g.log.WithFields(logrus.Fields{"key_version": unusedKey.keyVersionName}).WithError(err).Warn("error deleting unused GCP KMS key")
+			return trace.Wrap(err)
+		}
+	}
+	return nil
+}
+
+// kmsSigner implements the crypto.Signer interface
+type kmsSigner struct {
+	kmsClient kmsClientI
+	keyID     gcpKMSKeyID
+	public    crypto.PublicKey
+	clock     clockwork.Clock
+}
+
+func (g *gcpKMSKeyStore) newKmsSigner(ctx context.Context, keyID gcpKMSKeyID) (*kmsSigner, error) {
+	req := &kmspb.GetPublicKeyRequest{
+		Name: keyID.keyVersionName,
+	}
+	// Retry fetching the public key while the key is pending creation.
+	resp, err := retryWhilePending(ctx, g.clock, g.kmsClient.GetPublicKey, req)
+	if err != nil {
+		return nil, trace.Wrap(err, "unexpected error fetching public key")
+	}
+
+	block, _ := pem.Decode([]byte(resp.Pem))
+	if block == nil {
+		return nil, trace.BadParameter("GCP KMS key %s has invalid public key PEM", keyID.keyVersionName)
+	}
+	pub, err := x509.ParsePKIXPublicKey(block.Bytes)
+	if err != nil {
+		return nil, trace.Wrap(err, "unexpected error parsing public key pem")
+	}
+
+	return &kmsSigner{
+		kmsClient: g.kmsClient,
+		clock:     g.clock,
+		keyID:     keyID,
+		public:    pub,
+	}, nil
+}
+
+func (s *kmsSigner) Public() crypto.PublicKey {
+	return s.public
+}
+
+func (s *kmsSigner) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) (signature []byte, err error) {
```

Comment:
Why not plumb context through here?

Reply:
This is part of the `crypto.Signer` interface, which makes things a bit awkward. Perhaps we should reuse the same context that was used as part of `generateRSA`/`getSigner`?

Reply:
My concern with that was that the signer may outlive the context passed to `getSigner`, but I can look into it

Reply:
it seems fine to pass along the context from `generateRSA`/`getSigner`, I've gone ahead and done that